<a href="https://colab.research.google.com/github/ajitjadhav10/UMBC/blob/main/DATA%20690%20NLP/DATA_690_NLP_HW_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Exercise 1

In [93]:
import pandas as pd
import requests

In [94]:
data= requests.get('https://raw.githubusercontent.com/ajitjadhav10/UMBC/main/DATA%20690%20NLP/Data%20files/Copy%20of%20Burbank.txt')
data_1 = data.text
print(data_1)

Airport task force to consider possible actions to abate noise issues Members of the Southern San Fernando Valley Airplane Noise Task Force will have a little over a month to develop recommendations about how to address aircraft noise issues affecting the south San Fernando Valley After five meetings and hearing from various neighborhood groups and experts, aviation consultant HMMH Inc., which has been the facilitator of the meetings, spent the majority of Wednesday night distilling all of the recommendations that have been brought forward so far and the feasibility of implementing them 
HMMH worked with Kevin Karpe, an aviation expert who worked as an air traffic controller at Hollywood Burbank and the Federal Aviation Administration Southern California Terminal Radar Approach Control, to analyze the departing flight procedures out of the Burbank airfield as well as Van Nuys Airport, both of which are at the center of the noise issues 
Since March 2017, residents in the south San Fer

In [95]:
from textblob import TextBlob
import nltk
nltk.download('punkt')


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [96]:
blob=TextBlob(data_1)

In [172]:
total_sent=0
positive_sent=0
negative_sent=0
neutral_sent=0

for sentence in blob.sentences:
  total_sent +=1
  if sentence.sentiment.polarity > 0:
    positive_sent +=1
  elif sentence.sentiment.polarity == 0:
    neutral_sent +=1
  elif sentence.sentiment.polarity < 0:
    negative_sent +=1

print('probablity of negative sentiment:',(negative_sent/total_sent))

probablity of negative sentiment: 0.07142857142857142


Exercise 2

In [98]:
testimonial = TextBlob(data_1)
testimonial.sentiment

Sentiment(polarity=0.09869334480780263, subjectivity=0.3790877796901893)

The Sentiment of the text is positive and the subjectivity score is 0.3790877796901893

Exercise 3

In [99]:
word_list=blob.words
print(word_list)

['\ufeffAirport', 'task', 'force', 'to', 'consider', 'possible', 'actions', 'to', 'abate', 'noise', 'issues', 'Members', 'of', 'the', 'Southern', 'San', 'Fernando', 'Valley', 'Airplane', 'Noise', 'Task', 'Force', 'will', 'have', 'a', 'little', 'over', 'a', 'month', 'to', 'develop', 'recommendations', 'about', 'how', 'to', 'address', 'aircraft', 'noise', 'issues', 'affecting', 'the', 'south', 'San', 'Fernando', 'Valley', 'After', 'five', 'meetings', 'and', 'hearing', 'from', 'various', 'neighborhood', 'groups', 'and', 'experts', 'aviation', 'consultant', 'HMMH', 'Inc', 'which', 'has', 'been', 'the', 'facilitator', 'of', 'the', 'meetings', 'spent', 'the', 'majority', 'of', 'Wednesday', 'night', 'distilling', 'all', 'of', 'the', 'recommendations', 'that', 'have', 'been', 'brought', 'forward', 'so', 'far', 'and', 'the', 'feasibility', 'of', 'implementing', 'them', 'HMMH', 'worked', 'with', 'Kevin', 'Karpe', 'an', 'aviation', 'expert', 'who', 'worked', 'as', 'an', 'air', 'traffic', 'control

In [101]:
from string import punctuation
from spacy.lang.en.stop_words import STOP_WORDS
import spacy

In [102]:
extra_words=list(STOP_WORDS)+list(punctuation)+['\n']
nlp = spacy.load("en_core_web_sm")

In [103]:
important_words={}


for word in word_list:
  if word.lower() not in extra_words:
    if word not in important_words.keys():
      important_words[word]=1
    else:
      important_words[word] +=1

In [104]:
import operator
sorted_d = dict( sorted(important_words.items(), key=operator.itemgetter(1),reverse=True))

In [176]:
#The words are ordered in the order of their importance based on their frequency

print(sorted_d)

{'FAA': 13, 'Burbank': 12, 'said': 12, 'Valley': 11, '’': 10, 'recommendations': 9, 'HMMH': 9, 'Hollywood': 9, 'Southern': 8, 'San': 8, 'planes': 8, 'flights': 8, 'noise': 7, 'Fernando': 7, 'Task': 7, 'Force': 7, 'groups': 7, 'departing': 7, 'flight': 7, 'traffic': 6, 'California': 6, 's': 6, 'Scholten': 6, 'paths': 6, 'issues': 5, 'south': 5, 'air': 5, 'procedures': 5, 'NextGen': 5, 'northbound': 5, 'turns': 5, 'controllers': 5, 'community': 5, 'solutions': 5, 'Van': 4, 'Nuys': 4, '2017': 4, '—': 4, '101': 4, 'Freeway': 4, 'City': 4, 'region': 4, 'air-traffic': 4, '“': 4, 'meeting': 4, 'routes': 4, 'task': 3, 'force': 3, 'possible': 3, 'Noise': 3, 'meetings': 3, 'aviation': 3, 'Airport': 3, 'residents': 3, 'TRACON': 3, '”': 3, 'recommended': 3, 'consulting': 3, 'firm': 3, 'suggested': 3, 'waypoints': 3, 'North': 3, 'representatives': 3, 'airliner': 3, 'jet': 3, 'share': 3, 'Members': 2, 'Airplane': 2, 'month': 2, 'address': 2, 'aircraft': 2, 'neighborhood': 2, 'consultant': 2, 'Inc': 

Exercise 4

In [157]:
data_yelp=pd.read_csv('https://raw.githubusercontent.com/ajitjadhav10/UMBC/main/DATA%20690%20NLP/Data%20files/yelp_labelled.txt', sep='\t')
data_imdb=pd.read_csv('https://raw.githubusercontent.com/ajitjadhav10/UMBC/main/DATA%20690%20NLP/Data%20files/imdb_labelled.txt',sep='\t')
data_amazon=pd.read_csv('https://raw.githubusercontent.com/ajitjadhav10/UMBC/main/DATA%20690%20NLP/Data%20files/amazon_cells_labelled.txt',sep='\t')

In [158]:
#Joining the three data frames
combined_col=[data_amazon, data_imdb, data_yelp]

In [159]:
for column_name in combined_col:
  column_name.columns=['Review','Label']

In [160]:
for column_name in combined_col:
  print(column_name.columns)

Index(['Review', 'Label'], dtype='object')
Index(['Review', 'Label'], dtype='object')
Index(['Review', 'Label'], dtype='object')


In [161]:
keys_1=['amazon','imdb','yelp']

In [162]:
comb_data = pd.concat(combined_col,keys=keys_1).reset_index()
comb_data.head()

level_0  level_1                                             Review  Label
0  amazon        0                        Good case, Excellent value.      1
1  amazon        1                             Great for the jawbone.      1
2  amazon        2  Tied to charger for conversations lasting more...      0
3  amazon        3                                  The mic is great.      1
4  amazon        4  I have to jiggle the plug to get it to line up...      0

In [163]:
comb_data=comb_data.drop('level_1',axis=1)

In [164]:
comb_data=comb_data.rename({'level_0':'Company'}, axis=1)
comb_data.head()

comp_source                                             Review  Label
0      amazon                        Good case, Excellent value.      1
1      amazon                             Great for the jawbone.      1
2      amazon  Tied to charger for conversations lasting more...      0
3      amazon                                  The mic is great.      1
4      amazon  I have to jiggle the plug to get it to line up...      0

In [165]:
comb_data.columns

Index(['comp_source', 'Review', 'Label'], dtype='object')

In [166]:
comb_data.isnull().sum()

comp_source    0
Review         0
Label          0
dtype: int64

In [167]:
comb_data.to_csv("Sent_Analysis_Dataset.csv", index=False)

In [168]:
df_1 = pd.read_csv('Sent_Analysis_Dataset.csv')
df_1

comp_source                                             Review  Label
0         amazon                        Good case, Excellent value.      1
1         amazon                             Great for the jawbone.      1
2         amazon  Tied to charger for conversations lasting more...      0
3         amazon                                  The mic is great.      1
4         amazon  I have to jiggle the plug to get it to line up...      0
...          ...                                                ...    ...
2740        yelp  I think food should have flavor and texture an...      0
2741        yelp                           Appetite instantly gone.      0
2742        yelp  Overall I was not impressed and would not go b...      0
2743        yelp  The whole experience was underwhelming, and I ...      0
2744        yelp  Then, as if I hadn't wasted enough of my life ...      0

[2745 rows x 3 columns]

In [169]:
df_1.columns

Index(['comp_source', 'Review', 'Label'], dtype='object')

In [118]:
import string
import spacy
from spacy.lang.en.stop_words import STOP_WORDS
from spacy.lang.en import English
nlp = spacy.load("en_core_web_sm")

In [119]:
punctuations = string.punctuation

In [120]:
punctuations

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [122]:
stop_words = list(STOP_WORDS)
stop_words

['whole',
 'about',
 'themselves',
 'you',
 'could',
 'not',
 'wherein',
 'within',
 'yourself',
 'call',
 'nor',
 '’d',
 'whoever',
 'wherever',
 'empty',
 'several',
 "'ll",
 'few',
 'anyone',
 'me',
 'if',
 'also',
 'all',
 'through',
 'were',
 '‘d',
 'nevertheless',
 'even',
 'say',
 'had',
 'anything',
 'noone',
 'whereby',
 'it',
 'across',
 "'ve",
 'nowhere',
 'toward',
 'further',
 'some',
 'regarding',
 'anyhow',
 'eight',
 "'s",
 '‘ll',
 'she',
 'both',
 'why',
 'others',
 'between',
 'my',
 'ours',
 'amongst',
 'should',
 'six',
 'hereby',
 'those',
 'hence',
 'unless',
 'yours',
 'someone',
 'these',
 'has',
 'them',
 'always',
 'give',
 'n‘t',
 'ourselves',
 'hundred',
 'side',
 'doing',
 'so',
 'rather',
 'have',
 'whom',
 'sometime',
 'see',
 'a',
 'seemed',
 'three',
 'top',
 'for',
 'except',
 'move',
 'last',
 'anywhere',
 'seems',
 'to',
 'put',
 'again',
 'amount',
 'here',
 'everywhere',
 'an',
 'as',
 'please',
 'out',
 'then',
 'can',
 'five',
 'neither',
 'towar

In [131]:
from nltk.tokenize import RegexpTokenizer
regex_tokenizer = RegexpTokenizer(r'[a-zA-Z]+')

tokenizing = RegexpTokenizer('\s+', gaps=True)
df_1['data_tokens'] = df_1['Review'].map(tokenizing.tokenize)

In [124]:
from sklearn.metrics import accuracy_score 
from sklearn.model_selection import train_test_split 
from sklearn.base import TransformerMixin
from sklearn.svm import LinearSVC 
from sklearn.pipeline import Pipeline 
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.svm import LinearSVC

In [125]:
X = df_1['Review'] # the features we want to analyze
ylabels = df_1['Label'] # the labels, or answers, we want to test against

X_train, X_test, y_train, y_test = train_test_split(X, ylabels, test_size=0.3,random_state = 0)

In [127]:
# Custom transformer using spaCy
class predictors(TransformerMixin):
    def transform(self, X, **transform_params):
        # Cleaning Text
        return [clean_text(text) for text in X]

    def fit(self, X, y=None, **fit_params):
        return self

    def get_params(self, deep=True):
        return {}

nltk.download('stopwords')
stopwords=nltk.corpus.stopwords.words('english')
def clean_text(text):
    edited_text=re.sub('\W'," ",text) #replace any sumbol with whitespace
    edited_text=re.sub("  "," ",edited_text) #replace double whitespace with single whitespace
    edited_text=edited_text.split(" ") #split the sentence into array of strings
    edited_text=" ".join([char for char in edited_text if char!= ""]) #remove any empty string from text
    edited_text=edited_text.lower() #lowercase
    edited_text=re.sub('\d+',"",edited_text) #Removing numerics
    edited_text=re.split('\W+',edited_text) #spliting based on whitespace or whitespaces
    edited_text=" ".join([word for word in edited_text if word not in stopwords]) #Snowball Stemmer
    return edited_text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [145]:
bow_vector = TfidfVectorizer(lowercase=True,stop_words='english',ngram_range = (1,1),tokenizer = regex_tokenizer.tokenize)
classifier = LinearSVC()

In [135]:
pipe = Pipeline([("cleaner", predictors()),('vectorizer', bow_vector),('classifier', classifier)])
pipe.fit(X_train,y_train)

Pipeline(steps=[('cleaner', <__main__.predictors object at 0x7f6598115550>),
                ('vectorizer',
                 TfidfVectorizer(stop_words='english',
                                 tokenizer=<bound method RegexpTokenizer.tokenize of RegexpTokenizer(pattern='[a-zA-Z]+', gaps=False, discard_empty=True, flags=<RegexFlag.UNICODE|DOTALL|MULTILINE: 56>)>)),
                ('classifier', LogisticRegression())])

In [143]:
prediction_result = pipe.predict(X_test)
prediction_result

array([0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0,
       1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1,
       0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1,
       0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0,

In [171]:
accuracy_1=accuracy_score(y_test,prediction_result)
accuracy_1

0.7924757281553398